framework/src/base/FEProblem.C

```diff
@@ -3837,6 +3875,9 @@ FEProblem::checkNonlinearConvergence(std::string &msg,
                                      const Real initial_residual_before_preset_bcs,
                                      const Real div_threshold)
 {
+  if (_has_exception)
```

Comment:
I agree that this doesn't make sense, but at this point, we are actually *computing* the reason, so I don't think we can check the reason.  Instead, we could call `SNESGetFunctionDomainError` (returns true if `SNESSetFunctionDomainError()` has been called) in our `petscNonlinearConverged()` function, and set the appropriate reason there.  Not sure that's actually necessary, as PETSc already seems to diverge on its own with this reason, but I will test it.